src/Analyzers.CodeFixes/CSharp/Refactorings/SimplifyNullCheckRefactoring.cs

```diff
@@ -44,6 +45,16 @@ internal static class SimplifyNullCheckRefactoring
             //RCS1084 UseCoalesceExpressionInsteadOfConditionalExpression
             newNode = nullCheck.Expression;
             coalesce = true;
+            
+            // If the types are polymorphic then the LHS of the null coalesce must be cast to the base type.
+            var newNodeType = semanticModel.GetTypeInfo(newNode).Type;
+            var whenNullType = semanticModel.GetTypeInfo(whenNull).Type;
+            var overallType = semanticModel.GetTypeInfo(conditionalExpression).ConvertedType;
+            if (overallType != null && newNodeType != null && !newNodeType.Equals(whenNullType))
+            {
+                var typeCastStr = overallType.ToMinimalDisplayString(semanticModel, overallType.Locations.First().SourceSpan.Start);
+                newNode = CastExpression(NullableType(ParseTypeName(typeCastStr)), newNode);
+            }
```

Comment:
In general, this code does what is should do but there are many small improvements.

I wrote entire block of code instead of listing each suggestion separately:

```
ITypeSymbol newNodeType = semanticModel.GetTypeSymbol(newNode);
ITypeSymbol whenNullType = semanticModel.GetTypeSymbol(whenNull);
ITypeSymbol overallType = semanticModel.GetTypeInfo(conditionalExpression).ConvertedType;

if (overallType?.SupportsExplicitDeclaration() == true
    && !SymbolEqualityComparer.Default.Equals(newNodeType, whenNullType))
{
    TypeSyntax castType = overallType.ToTypeSyntax().WithSimplifierAnnotation();

    if ((semanticModel.GetNullableContext(conditionalExpression.SpanStart) & NullableContext.AnnotationsEnabled) != 0)
        castType = NullableType(castType);

    newNode = CastExpression(
        castType,
        newNode.WithoutTrivia())
        .WithTriviaFrom(newNode);
}
```